src/pal/src/include/pal/sharedmemory.h

```diff
@@ -0,0 +1,247 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#ifndef _PAL_SHARED_MEMORY_H_
+#define _PAL_SHARED_MEMORY_H_
+
+#include "corunix.hpp"
+
+#ifndef static_assert_no_msg
+#define static_assert_no_msg( cond ) static_assert( cond, #cond )
+#endif // !static_assert_no_msg
+
+#ifndef _countof
+#define _countof(a) (sizeof(a) / sizeof(a[0]))
+#endif // !_countof
+
+#ifndef PAGE_SIZE
+#define PAGE_SIZE (0x1000)
+#endif // !PAGE_SIZE
+
+// - Global shared memory files go in:
+//     /tmp/coreclr/shm/global/<fileName>
+// - Session-scoped shared memory files go in:
+//     /tmp/coreclr/shm/session<sessionId>/<fileName>
+// - Lock files associated with global shared memory files go in:
+//     /tmp/coreclr/lockfiles/global/<fileName>
+// - Lock files associated with session-scoped shared memory files go in:
+//     /tmp/coreclr/lockfiles/session<sessionId>/<fileName>
+
+#define SHARED_MEMORY_MAX_FILE_NAME_CHAR_COUNT (_MAX_FNAME - 1)
+#define SHARED_MEMORY_MAX_NAME_CHAR_COUNT (_countof("Global\\") - 1 + SHARED_MEMORY_MAX_FILE_NAME_CHAR_COUNT)
+
+#define SHARED_MEMORY_TEMP_DIRECTORY_PATH "/tmp"
+#define SHARED_MEMORY_RUNTIME_TEMP_DIRECTORY_PATH "/tmp/coreclr"
+
+#define SHARED_MEMORY_SHARED_MEMORY_DIRECTORY_PATH "/tmp/coreclr/shm"
+#define SHARED_MEMORY_LOCK_FILES_DIRECTORY_PATH "/tmp/coreclr/lockfiles"
```

Comment:
On second thought, I'm putting specific permissions on /tmp/coreclr and all files/directories below so that they have read/write permissions for all users. I'm not sure if other pieces that use /tmp/dotnet do this. Although I make an attempt to update permissions on the directories if they already exist, that may fail if the user doesn't have the necessary permissions to do so. I'll leave the root as /tmp/coreclr for now.